discovery/gossiper.go

```diff
@@ -2345,70 +2313,234 @@ func (d *AuthenticatedGossiper) fetchNodeAnn(
+					msg, compressedPeerPubKey,
+				)
+				if err != nil {
+					log.Errorf("Unable to remove stale %v "+
+						"message for channel=%v with "+
+						"peer %x: %v", msg.MsgType(),
+						shortChanID,
+						compressedPeerPubKey, err)
+				}
 
-			select {
-			case <-peerChan:
-				// Retry sending.
-				log.Infof("Peer %x reconnected. Retry sending"+
-					" AnnounceSignatures.",
-					remotePeer.SerializeCompressed())
+				continue
+			}
 
-			case <-d.quit:
-				log.Infof("Gossiper shutting down, did not " +
-					"send AnnounceSignatures.")
-				return
+			// TODO(wilmer): is it worth checking if they're still
+			// online at this point?
+			if err := peer.SendMessage(false, msg); err != nil {
+				log.Errorf("Unable to send %v message for "+
+					"channel=%v to %x: %v", msg.MsgType(),
+					shortChanID, compressedPeerPubKey, err)
+				continue
 			}
+
+			log.Debugf("Successfully sent %v message for "+
+				"channel=%v with peer=%x upon reconnection",
+				msg.MsgType(), shortChanID,
+				compressedPeerPubKey)
 		}
 
-		log.Infof("Sent channel announcement proof to remote peer: %x",
-			remotePeer.SerializeCompressed())
-	}()
+		// Once the messages are sent, we can request a notification for
+		// when the peer disconnects. We do this to prevent the loop
+		// from proceeding and avoid repeatedly requesting online
+		// notifications for the peer.
+		offlineChan := d.cfg.NotifyWhenOffline(compressedPeerPubKey)
+		select {
+		case <-offlineChan:
+		case <-d.quit:
+			return
+		}
 
-	// This method returns after the message has been added to the database,
-	// such that the caller don't have to wait until the message is actually
-	// delivered, but can be assured that it will be delivered eventually
-	// when this method returns.
-	return nil
+		// Finally, now that we've been signaled that they're offline,
+		// we can remove them from our cache and proceed to request
+		// another notification upon reconnection.
+		d.cachedPeersMtx.Lock()
+		delete(d.cachedPeers, compressedPeerPubKey)
+		d.cachedPeersMtx.Unlock()
+	}
+}
+
+// isStaleMsg determines whether a message retrieved from the backing
+// MessageStore is seen as stale by the current graph.
+func (d *AuthenticatedGossiper) isStaleMsg(msg lnwire.Message) bool {
+	switch msg := msg.(type) {
+	case *lnwire.AnnounceSignatures:
+		chanInfo, _, _, err := d.cfg.Router.GetChannelByID(
+			msg.ShortChannelID,
+		)
+		// If the channel cannot be found, it is most likely a leftover
+		// message for a channel that was closed, so we can consider it
+		// stale.
+		if err == channeldb.ErrEdgeNotFound {
+			return true
+		}
+		if err != nil {
+			log.Debugf("Unable to retrieve channel=%v from graph: "+
+				"%v", err)
+		}
+
+		// If the proof exists in the graph, then we have successfully
+		// received the remote proof and assembled the full proof, so we
+		// can safely delete the local proof from the database.
+		return chanInfo.AuthProof != nil
```

Comment:
This'll panic if we get any other error than `ErrEdgeNotFound`.

Reply:
Fixed.